test/e2e/throughput_test.go

```diff
@@ -0,0 +1,201 @@
+	}
+
+	if _, err := test.CreateRoute(logger, clients, helloWorldNames); err != nil {
+		t.Fatalf("Failed to create Route: %v", err)
+	}
+
+	test.CleanupOnInterrupt(func() { TearDown(clients, helloWorldNames, logger) }, logger)
+	defer TearDown(clients, helloWorldNames, logger)
+
+	revision, err := test.WaitForConfigLatestRevision(clients, helloWorldNames)
+	if err != nil {
+		t.Fatalf("Configuration %s was not updated with the new revision: %v", helloWorldNames.Config, err)
+	}
+
+	logger.Info("When the Route reports as Ready, everything should be ready.")
+	if err := test.WaitForRouteState(clients.ServingClient, helloWorldNames.Route, test.IsRouteReady, "RouteIsReady"); err != nil {
+		t.Fatalf("The Route %s was not marked as Ready to serve traffic: %v", helloWorldNames.Route, err)
+	}
+
+	helloWorldRoute, err := clients.ServingClient.Routes.Get(helloWorldNames.Route, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get Route %q of helloworld app: %v", helloWorldNames.Route, err)
+	}
+
+	domain := helloWorldRoute.Status.Domain
+
+	deploymentName := revision + "-deployment"
+
+	logger.Info("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		fmt.Errorf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	// Hit the service with concurrent requests to provoke the start from 0,
+	// the requests should go through activator, we make sure we receive only 200 responses.
+
+	logger.Info("Waiting for endpoint to serve request")
+
+	endpoint, err := spoof.GetServiceEndpoint(clients.KubeClient.Kube)
+	url := fmt.Sprintf("http://%s/?timeout=100", *endpoint)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, logger, domain, test.ServingFlags.ResolvableDomain)
+
+	totalRequests := 1000
+	responseChannel := make(chan *spoof.Response, totalRequests)
+	timeout := 30 * time.Second
+	roundTrip := roundTrip(client, url)
+
+	sendRequests(roundTrip, 1000, totalRequests, responseChannel, timeout, logger, t)
+
+	collectResponses(responseChannel, totalRequests, timeout, t)
+}
+
+func sendRequests(roundTrip func() (*spoof.Response, error), concurrency, totalRequests int, resChannel chan *spoof.Response, timeout time.Duration, logger *logging.BaseLogger, t *testing.T) int32 {
+	var (
+		group     errgroup.Group
+		responses int32
+	)
+	// The eventual number of sent requests is mod(concurrency).
+	runs := totalRequests / concurrency
+	tickChan := make(chan struct{}, runs)
+	doneChan := make(chan struct{})
+	timeoutChan := time.After(timeout)
+	errChan := make(chan error)
+	allDoneChan := make(chan struct{})
+
+	// Send a tick for each batch of requests.
+	go func() {
+		for i := 0; i < runs; i++ {
+			tickChan <- struct{}{}
+			<-doneChan
+		}
+		allDoneChan <- struct{}{}
+	}()
+
+	// Run a batch of requests.
+	go func() {
```

Comment:
We do have a thread pool, which seems to do things you do here. Perhaps try to utilize it?

Reply:
I removed that peace of code.